mobile/src/employee-details/event-action-provider.ts

```diff
@@ -0,0 +1,120 @@
+export interface EventAction {
+    readonly handler: () => void;
+    readonly name: string;
+}
+
+//============================================================================
+export interface EventActionsContainer {
+    readonly event: CalendarEvent;
+    readonly employee: Employee;
+    readonly positiveAction?: EventAction;
+    readonly negativeAction?: EventAction;
+}
+
+//============================================================================
+export class EventActionProvider {
+    //----------------------------------------------------------------------------
+    constructor(private readonly userId: EmployeeId,
+                private readonly eventSetStatus: (employeeId: EmployeeId, calendarEvent: CalendarEvent, status: CalendarEventStatus) => void,
+                private readonly eventApprove: (employeeId: EmployeeId, calendarEvent: CalendarEvent) => void) {
+    }
+
+    //----------------------------------------------------------------------------
+    public getEventActions(events: CalendarEvent[],
+                           employee: Employee,
+                           permissions: UserEmployeePermissions,
+                           approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer[] {
+
+        const eventContainers: EventActionsContainer[] = [];
+
+        for (const event of events) {
+            eventContainers.push(this.evaluateEvent(event, employee, permissions, approvals));
+        }
+
+        return eventContainers;
+    }
+
+    //----------------------------------------------------------------------------
+    public getRequestActions(events: CalendarEvent[],
+                             employee: Employee,
+                             approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer[] {
+
+        const permissions = new UserEmployeePermissions();
+        permissions.permissionsNames = Set([Permission.approveCalendarEvents, Permission.rejectCalendarEvents]);
+        return this.getEventActions(events, employee, permissions, approvals);
+    }
+
+    //----------------------------------------------------------------------------
+    private evaluateEvent(event: CalendarEvent,
+                          employee: Employee,
+                          permissions: UserEmployeePermissions,
+                          approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer {
+        let positiveAction: Optional<EventAction>;
+        let negativeAction: Optional<EventAction>;
+
+        const isOwn = this.userId === employee.employeeId;
+
+        if (event.status === CalendarEventStatus.Requested) {
+            if (permissions.has(Permission.approveCalendarEvents)) {
+                if (!this.approvedByMe(event, approvals)) {
+                    positiveAction = {
+                        handler: () => this.eventApprove(employee.employeeId, event),
+                        name: 'approve',
```

Comment:
suggestion to rename it to 'verb', since you use this 'name' in messages only. Or, better, to have a enum for names here and 'name -> verb/message' mapping in the component. Atm it looks like encapsulation violation (I have no idea editing this code that it'll change displayed messages).

Reply:
Done. 52173071967f3ef8d6fc6bb171f1b328eb137b56
Encapsulation is restored. -)